### PR TITLE
Refactor get program definition to throw an exception when not found

### PR DESCRIPTION
**Pull request by hmhsu**
### Description
After this change `ProgramService#getProgramDefinitionAsync` may throw exceptions after the future returned completes. The exception can propagate if the future is concatenated with other completable stages.

### Checklist
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
[Fixes] #<issue_number>

### CODE REVIEW

**Review by bion — APPROVED** [review]

**Review by kendrickt — COMMENTED** [review]
(blocking) the `ApplicantProgramsController.edit` uses `ApplicantService` and needs to use `exceptionally`. There may be others.

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java [hunk 1]
```diff
@@ -17,6 +17,8 @@
    * @return a {@link ReadOnlyApplicantProgramService} that reflects the updates, which may have
    *     invalid data with errors associated with it. If the service cannot perform the update, an
    *     {@link ErrorAnd} is returned in the error state.
+   *     <p>A ProgramNotFoundException may be thrown when the future completes if the programId does
```
[the remark this answers is not in the document]
Comment (hmhsu) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantService.java:20]: Unfortunately that doesn't compile as we are not actually throwing any exception with `throws`.

File: universal-application-tool-0.0.1/app/controllers/admin/AdminProgramBlocksController.java [hunk 2]
```diff
@@ -74,24 +72,22 @@ public Result create(long programId) {
...
   public Result edit(Request request, long programId, long blockId) {
-    Optional<ProgramDefinition> programMaybe = programService.getProgramDefinition(programId);
+    try {
+      ProgramDefinition program = programService.getProgramDefinition(programId);
+      Optional<BlockDefinition> blockMaybe = program.getBlockDefinition(blockId);
 
-    if (programMaybe.isEmpty()) {
-      return notFound(String.format("Program ID %d not found.", programId));
-    }
+      if (blockMaybe.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (hmhsu) [reply universal-application-tool-0.0.1/app/controllers/admin/AdminProgramBlocksController.java:79]: Sounds good!

[the remark this answers is not in the document]
Comment (hmhsu) [reply universal-application-tool-0.0.1/app/controllers/admin/AdminProgramBlocksController.java]: Done.

File: universal-application-tool-0.0.1/test/services/program/ProgramServiceImplTest.java [hunk 3]
```diff
@@ -224,17 +224,15 @@ public void updateProgram_constructsQuestionDefinitions() throws Exception {
 
   @Test
   public void getProgramDefinition_canGetANewProgram() throws Exception {
```
Comment (kendrickt) [inline universal-application-tool-0.0.1/test/services/program/ProgramServiceImplTest.java:226]: (blocking) The test name `canGetANewProgram` kind of implies that `ps.createProgramDefinition` is part of the test. I don't think there's any real value there. Can you change the test name to be more generic?

```suggestion
  public void getProgramDefinition() throws Exception {
```

Comment (hmhsu) [reply universal-application-tool-0.0.1/test/services/program/ProgramServiceImplTest.java]: Done.

File: universal-application-tool-0.0.1/app/controllers/admin/AdminProgramBlocksController.java [hunk 4]
```diff
@@ -74,19 +73,19 @@ public Result create(long programId) {
...
-      }
-
-      BlockDefinition block = blockMaybe.get();
       ReadOnlyQuestionService roQuestionService =
           questionService.getReadOnlyQuestionService().toCompletableFuture().join();
 
       return ok(editView.render(request, program, block, roQuestionService.getAllQuestions()));
     } catch (ProgramNotFoundException e) {
       return notFound(e.toString());
+    } catch (ProgramBlockNotFoundException e) {
```
[the remark this answers is not in the document]
Comment (hmhsu) [reply universal-application-tool-0.0.1/app/controllers/admin/AdminProgramBlocksController.java]: Done.

File: universal-application-tool-0.0.1/app/controllers/admin/AdminProgramBlocksController.java [hunk 5]
```diff
@@ -74,19 +73,19 @@ public Result create(long programId) {
...
-      if (blockMaybe.isEmpty()) {
-        return notFound(String.format("Block ID %d not found for Program %d", blockId, programId));
-      }
-
-      BlockDefinition block = blockMaybe.get();
       ReadOnlyQuestionService roQuestionService =
           questionService.getReadOnlyQuestionService().toCompletableFuture().join();
 
       return ok(editView.render(request, program, block, roQuestionService.getAllQuestions()));
     } catch (ProgramNotFoundException e) {
```
[the remark this answers is not in the document]
Comment (hmhsu) [reply universal-application-tool-0.0.1/app/controllers/admin/AdminProgramBlocksController.java]: Done.

**Comment by hmhsu** [comment]
> (blocking) the `ApplicantProgramsController.edit` uses `ApplicantService` and needs to use `exceptionally`. There may be others.

Working on this :)

File: universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramBlocksController.java [hunk 6]
```diff
@@ -70,7 +72,18 @@ public ApplicantProgramBlocksController(
                 return notFound();
               }
             },
-            httpExecutionContext.current());
+            httpExecutionContext.current())
+        .exceptionally(
+            ex -> {
```
Comment (kendrickt) [inline universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramBlocksController.java:77]: (optional): should this be refactored into a method that looks like (loosely)

```
public Result handleExceptions(ImmutableSet<Class<? extends Exception>> okExceptionClasses, Throwable exception) {
  for (Class okExceptionClass : okExceptionClasses) {
    if (okExceptionClass.isInstance(exception)) {
      return notFound(exception.toString());    
    }
  }
 throw new RuntimeException(exception);
}
```

My method and arg names are not great.

I can imagine some helper methods to construct the `ImmutableSet` from the bottom up, so we don't have to manage things so manually.

**Review by kendrickt — APPROVED** [review]

File: universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramBlocksController.java (hunk 6 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramBlocksController.java:77]: the `if` case probably log the error and return `notFound("Unable to process this request.")`

Comment (hmhsu) [reply universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramBlocksController.java:77]: That sounds like a great idea. Opened #487 and would like to merge this PR first.

**Pull request merged**